test/src/test/java/org/corfudb/integration/MergeTestListener.java

```diff
@@ -0,0 +1,204 @@
+import static org.corfudb.integration.MergeTestListenerIT.GROUPA_TABLE_PREFIX;
+import static org.corfudb.integration.MergeTestListenerIT.LOCAL_TABLE_PREFIX;
+import static org.corfudb.integration.MergeTestListenerIT.MERGED_TABLE_NAME;
+import static org.junit.Assert.fail;
+
+/**
+ * This class is for LogicalGroup E2E IT
+ *
+ *  (1) Open a merged table that store all the existing data and keys,
+ *      add keys to a set that store all the existing keys
+ *  (2) Start snapshot sync
+ *  (3) Process updates for snapshot sync, perform the diff by using
+ *      the set in (1) when processing replicated tables
+ *  (4) Snapshot sync ends, add the keys to the existing keys' set
+ *      and apply the keys to merged table
+ *  (5) Process updates for log entry sync
+ */
+@Slf4j
+public class MergeTestListener extends LogReplicationListener{
+
+    public CorfuRuntime runtime;
+
+    private CorfuStore corfuStoreSink;
+
+    private final String namespace = "LR-Test";
+
+    // Updates received through streaming
+    @Getter
+    public static LinkedList<CorfuStreamEntry> updates = new LinkedList<>();
+
+    public Set<Sample.StringKey> deletedKeys = new HashSet<>();
+
+    public Map<Sample.StringKey, CorfuStoreEntry> addedKeys = new HashMap<>();
+
+    public Set<Sample.StringKey> existingKeysInMergedTable = new HashSet<>();
+
+
+    public static List<CorfuStoreEntry<Sample.StringKey, SampleSchema.SampleGroupMsgA, Message>> entries
+            = new ArrayList<>();
+
+    private final int numTableForGroupA;
+
+    private final int numTableForLocalTables;
+
+    MergeTestListener(CorfuStore corfuStoreSink, String namespace, int numTableForGroupA, int numTableForLocalTables) {
+        super(corfuStoreSink, namespace);
+        this.corfuStoreSink = corfuStoreSink;
+        this.numTableForGroupA = numTableForGroupA;
+        this.numTableForLocalTables = numTableForLocalTables;
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        log.error("Error on listener", throwable);
+        fail("onError for LRTestListener: " + throwable.toString());
+    }
+
+    /**
+     * Invoke when snapshot sync starts
+     */
+    @Override
+    protected void onSnapshotSyncStart() {
+        log.info("Snapshot sync started");
+    }
+
+    /**
+     * Apply addedKeys and deletedKeys to merged table, add keys to
+     * existingKeysInMergedTable from addedKeys and deletedKeys.
+     */
+    @Override
+    protected void onSnapshotSyncComplete() {
+        try (TxnContext tx = corfuStoreSink.txn(NAMESPACE)){
+            existingKeysInMergedTable.addAll(addedKeys.keySet());
+            addedKeys.forEach((addedKey, e) -> {
+                tx.putRecord(tx.getTable(MERGED_TABLE_NAME), addedKey,
+                        SampleSchema.SampleMergedTable.newBuilder().setPayload(e.getPayload().toString().replaceAll("[^0-9]", "")).build(), e.getMetadata());
+            });
```

Comment:
why the replaceAll needed here?

Reply:
It is a way to pass the verification of the IT. 

For `CorfuStoreEntry`(See `performFullSync` method), we do not need `replaceAll` since `entry.getPayload().getPayload()` returns payload's number in String type(For example, if `entry.getPayload()` returns payload: "597", then `entry.getPayload().getPayload()` returns "597"). 

For `CorfuStreamEntry`(See 'processUpdatesInLogEntrySync'), it seems that we couldn't do `e.getPayload().getPayload()` to get payload's number in String type. Instead, we can do `e.getPayload().toString()` to get the payload in String type(For example, if `e.getPayload()` returns payload: "597", then `e.getPayload().getPayload()` returns "payload: "597""), then invoke `replaceAll` to get the payload number.

Please let me know if there is a cleaner way to do this.